apps/hotel/src/domain/GuestStay.ts

```diff
@@ -0,0 +1,201 @@
+        at: OffsetDateTime
+        residualBalance: number
+        groupId: GroupCheckoutId
+      }
+
+  export const initial: State = { type: "Active", balance: 0, charges: [], payments: [] }
+
+  export const fold = createFold<Events.Event, State>({
+    CheckedIn(state, { at }) {
+      if (state.type !== "Active") throw new Error("Invalid state transition")
+      state.checkedInAt = at
+    },
+    Charged(state, { chargeId, amount }) {
+      if (state.type !== "Active") throw new Error("Invalid state transition")
+      state.balance += amount
+      state.charges.push(chargeId)
+    },
+    Paid(state, { paymentId, amount }) {
+      if (state.type !== "Active") throw new Error("Invalid state transition")
+      state.balance -= amount
+      state.payments.push(paymentId)
+    },
+    CheckedOut(state) {
+      if (state.type !== "Active") throw new Error("Invalid state transition")
+      return { type: "Closed" }
+    },
+    TransferredToGroup(state, { at, groupId, residualBalance }) {
+      if (state.type !== "Active") throw new Error("Invalid state transition")
+      return { type: "TransferredToGroup", at, residualBalance, groupId }
+    },
+  })
+}
+
+export namespace Decide {
+  import State = Fold.State
+  import Event = Events.Event
+
+  type Decision = (state: State) => Event[]
+  type DecisionResult<T> = (state: State) => [T, Event[]]
+
+  export const checkIn =
+    (at: OffsetDateTime): Decision =>
+    (state) => {
+      if (state.type !== "Active") throw new Error("Invalid checkin")
+      if (state.checkedInAt) return []
+      return [Events.CheckedIn({ at })]
+    }
+
+  export const charge =
+    (at: OffsetDateTime, chargeId: ChargeId, amount: number): Decision =>
+    (state) => {
+      if (state.type !== "Active") throw new Error("Invalid charge")
+      if (state.charges.includes(chargeId)) return []
+      return [Events.Charged({ chargeId, at, amount })]
+    }
+
+  export const payment =
+    (at: OffsetDateTime, paymentId: PaymentId, amount: number): Decision =>
+    (state) => {
+      if (state.type !== "Active") throw new Error("Invalid payment")
+      if (state.payments.includes(paymentId)) return []
+      return [Events.Paid({ paymentId, at, amount })]
+    }
+
+  type CheckoutResult =
+    | { type: "Ok" }
+    | { type: "AlreadyCheckedOut" }
+    | { type: "BalanceOutstanding"; balance: number }
+
+  export const checkOut =
+    (at: OffsetDateTime): DecisionResult<CheckoutResult> =>
+    (state) => {
+      switch (state.type) {
+        case "Closed":
+          return [{ type: "Ok" }, []]
+        case "TransferredToGroup":
+          return [{ type: "AlreadyCheckedOut" }, []]
+        case "Active":
+          return state.balance === 0
+            ? [{ type: "Ok" }, [Events.CheckedOut({ at })]]
+            : [{ type: "BalanceOutstanding", balance: state.balance }, []]
+      }
+    }
+
+  type GroupCheckoutResult = { type: "Ok"; residualBalance: number } | { type: "AlreadyCheckedOut" }
```

Comment:
the on time you want prettier to do its goddam job!

Reply:
currently have https://github.com/fsprojects/FSharp.AWS.DynamoDB/pull/73  wandering around my head...

Reply:
Yeah, I'm getting tired of tilting at that windmill